File: main_project/tripod-new/src/cpp/configutil/ConfigManager.cpp
#include <iostream>
#include <boost/pointer_cast.hpp>

#include <LogWrapper.h>

#include "ZkManager.h"
#include "ZkManagerFactory.h"
#include "ConfigClient.h"
#include "ConfigData.h"

#include "ConfigManager.h"

namespace com {
namespace renren {
namespace tripod {

BufferContainer::BufferContainer() {
    buflen = sizeof(buffer);
    memset(buffer, 0, buflen);
}

ConfigManager::ConfigManager() {
}

bool ConfigManager::initialize(const std::string& zkAddress) {
    MCE_INFO("ConfigManager::initialize zkAddress " << zkAddress);
    zm_ = ZkManagerFactory::getInstance()->getZkManager(zkAddress);
    if( zm_ == NULL ) {
        MCE_WARN("ConfigManager::initialize zkManager initialize failed!");
        return false;
    }

    return updateConfig(zm_->getZk());
}

bool ConfigManager::updateConfig(zhandle_t* zk) {
    MCE_INFO("ConfigManager::updateConfig");

    ConfigDataPtr newConfigData = getConfigData();

    bool updateFlag = false;

    bool sucFlag = init(newConfigData, zk, updateFlag); 

    if(updateFlag && sucFlag) {
        MCE_INFO("ConfigManager::updateConfig set newconfig");
        boost::upgrade_lock<boost::shared_mutex> lock(lock_);
        boost::upgrade_to_unique_lock<boost::shared_mutex> uniqueLock(lock);
        configData_ = newConfigData;
    }

    if(updateFlag) {
        MCE_INFO("ConfigManager::updateConfig update clients");
        boost::shared_lock<boost::shared_mutex> lock(clientsLock_);
        for(std::vector<ConfigClientPtr>::const_iterator it = clients_.begin(); it != clients_.end(); ++it) {
            (*it)->notifyHandle();
        }
    }
    
    MCE_INFO("ConfigManager::updateConfig end sucFlag " << sucFlag << " updateFlag " << updateFlag);

    return sucFlag;
}

void ConfigManager::registerClient(const ConfigClientPtr& cc) {
    MCE_INFO("ConfigManager::registerClient");
    boost::upgrade_lock<boost::shared_mutex> lock(clientsLock_);
    boost::upgrade_to_unique_lock<boost::shared_mutex> uniqueLock(lock);
    clients_.push_back(cc);
}

ConfigDataPtr ConfigManager::getConfigData() {
    MCE_INFO("ConfigManager::getConfigData");
    boost::shared_lock<boost::shared_mutex> lock(lock_);
    if(configData_ != NULL) {
        return configData_->getData();
    }else
    {
        MCE_WARN("ConfigManager::getConfigData configdata_ is NULL");
        return ConfigDataPtr();
    }
}

bool ConfigManager::init(ConfigDataPtr& config, zhandle_t* zk, bool& updateFlag) {
    MCE_INFO("ConfigManager::init");

    updateFlag = false;

    if(config == NULL) {
        MCE_INFO("ConfigManager::init first time config null");
        ConfigDataPtr managerConfig(new ConfigData);
        managerConfig->name_ = TRIPOD_MANAGER;
        ConfigDataPtr producerConfig(new ConfigData);
        producerConfig->name_ = TRIPOD_PRODUCER;
        ConfigDataPtr cacheConfig(new ConfigData);
        cacheConfig->name_ = TRIPOD_CACHE;
        updateFlag = true;
        bool tempFlag = false;
        bool sucFlag = updateManagerConfig(managerConfig, zk, false, tempFlag) && updateProducerConfig(producerConfig, zk, false, tempFlag) && updateCacheConfig(cacheConfig, zk, false, tempFlag);
        if(sucFlag) {
            config = ConfigDataPtr(new ConfigData);
            config->childrenDatas_.insert(make_pair(TRIPOD_MANAGER, managerConfig));
            config->childrenDatas_.insert(make_pair(TRIPOD_PRODUCER, producerConfig));
            config->childrenDatas_.insert(make_pair(TRIPOD_CACHE, cacheConfig));
        }
        MCE_INFO("ConfigManager::init sucFlag " << sucFlag);
        return sucFlag;
    }

    std::map<std::string, ConfigDataPtr>::const_iterator findIt = config->childrenDatas_.find(TRIPOD_MANAGER);

    if(findIt == config->childrenDatas_.end()) {
        MCE_WARN("ConfigManager::init no cachemanager in newconfig");
        return false;
    }

    bool updateManagerFlag = false;
    bool sucManagerFlag = updateManagerConfig(findIt->second, zk, true, updateManagerFlag);

    if(sucManagerFlag == false) {
        MCE_WARN("ConfigManager::init updateManagerConfig failed");
        return false;
    }

    findIt = config->childrenDatas_.find(TRIPOD_PRODUCER);

    if(findIt == config->childrenDatas_.end()) {
        MCE_WARN("ConfigManager::init no producermanager in newconfig");
        return false;
    }

    bool updateProducerFlag = false;
    bool sucProducerFlag = updateProducerConfig(findIt->second, zk, true, updateProducerFlag);

    if(sucProducerFlag == false) {
        MCE_WARN("ConfigManager::init updateProducerConfig failed");
        return false;
    }

    findIt = config->childrenDatas_.find(TRIPOD_CACHE);

    if(findIt == config->childrenDatas_.end()) {
        MCE_WARN("ConfigManager::init no cache in newconfig");
        return false;
    }

    bool updateCacheFlag = false;
    bool sucCacheFlag = updateCacheConfig(findIt->second, zk, true, updateCacheFlag);

    if(sucCacheFlag == false) {
        MCE_WARN("ConfigManager::init updateCacheConfig failed");
        return false;
    }

    updateFlag = updateCacheFlag | updateProducerFlag | updateManagerFlag;
    MCE_INFO("ConfigManager::init updateCacheFlag " << updateCacheFlag 
            << " updateManagerFlag " << updateManagerFlag 
            << " updateProducerFlag " << updateProducerFlag 
            << " updateFlag " << updateFlag);
    MCE_INFO("ConfigManager::init sucCacheFlag " << sucCacheFlag
            << " sucProducerFlag " << sucProducerFlag
            << " sucManagerFlag " << sucManagerFlag);

    return sucCacheFlag && sucProducerFlag && sucManagerFlag; 
}

bool ConfigManager::updateManagerConfig(const ConfigDataPtr& config, zhandle_t* zk, bool checkVersion, bool& updateFlag) {
    MCE_INFO("ConfigManager::updateManagerConfig");

    updateFlag =false;

    try
    {
        int watch = 1;

        BufferContainerPtr bc( new BufferContainer ); 

        std::ostringstream oss;
        oss << "/"<<TRIPOD_MANAGER;
        int rc = 0;
        rc = zoo_exists(zk, oss.str().c_str(), watch, 0);

        if (rc != ZOK) {
            MCE_WARN( "ConfigManager::updateManagerConfig zoo_exists failed because "<<zerror(rc)<< __FILE__<<__LINE__ );
            return false;
        }

        rc = zoo_get(zk, oss.str().c_str(), watch, bc->buffer, &(bc->buflen), 0);

        if (rc != ZOK) {
            MCE_WARN( "ConfigManager::updateManagerConfig zoo_get failed because "<<zerror(rc)<< __FILE__<<__LINE__ );
            return false;
        }
        std::string version(bc->buffer, bc->buffer + bc->buflen);
        if(checkVersion) {
            if(config->version_ == version) {
                MCE_INFO("ConfigManager::updateManagerConfig version is same");
                return true;
            }
        }

        config->name_ = TRIPOD_MANAGER;
        config->version_ = version;

        bool sucChildrenFlag = getChildrenDatas(config, zk, oss);

        if(!sucChildrenFlag) {
            MCE_WARN("ConfigManager::updateManagerConfig getChildrenDatas failed");
            return false;
        }
    }catch(...)
    {
        MCE_WARN( "ConfigManager::updateManagerConfig can not init config "<< __FILE__<<__LINE__ );
        return false;
    }

    updateFlag = true;
    return true;
}

bool ConfigManager::updateProducerConfig(const ConfigDataPtr& config, zhandle_t* zk, bool checkVersion, bool& updateFlag) {
    MCE_INFO("ConfigManager::updateProducerConfig");

    updateFlag =false;

    try
    {
        int watch = 1;

        BufferContainerPtr bc( new BufferContainer ); 

        std::ostringstream oss;
        oss << "/"<<TRIPOD_PRODUCER<<"/"<<TRIPOD_PRODUCER_CC;
        
        int rc = zoo_exists(zk, oss.str().c_str(), watch, 0);

        if (rc != ZOK) {
            MCE_WARN( "ConfigManager::updateProducerConfig zoo_exists failed because "<<zerror(rc)<< __FILE__<<__LINE__ );
            updateFlag =false;
            return false;
        }

        rc = zoo_get(zk, oss.str().c_str(), watch, bc->buffer, &(bc->buflen), 0);

        if (rc != ZOK) {
            MCE_WARN( "ConfigManager::updateProducerConfig zoo_get failed because "<<zerror(rc)<< __FILE__<<__LINE__ );
            updateFlag =false;
            return false;
        }

        std::string ccVersion(bc->buffer, bc->buffer + bc->buflen);
        if(checkVersion) {
            MCE_INFO("ConfigManager::updateProducerConfig cc need check version");
            std::map<std::string, ConfigDataPtr>::const_iterator findIt = config->childrenDatas_.find(TRIPOD_PRODUCER_CC);
            if(findIt == config->childrenDatas_.end()) {
                MCE_WARN("ConfigManager::updateProducerConfig no cc");
                updateFlag =false;
                return false;
            }
            if(findIt->second->version_ != ccVersion) {
                updateFlag = true;
                findIt->second->version_ = ccVersion;
                bool sucCCFlag = getChildrenDatas(findIt->second, zk, oss);
                if(sucCCFlag == false) {
                    MCE_WARN("ConfigManager::updateProducerConfig get cc childrendatas failed");
                    updateFlag =false;
                    return false;
                }
            }
        }else {
            MCE_INFO("ConfigManager::updateProducerConfig cc do not need check version");
            ConfigDataPtr newConfig(new ConfigData);
            newConfig->version_ = ccVersion;
            bool sucCCFlag = getChildrenDatas(newConfig, zk, oss);
            if(sucCCFlag == false) {
                MCE_WARN("ConfigManager::updateProducerConfig get cc childrendatas failed");
                updateFlag =false;
                return false;
            }
            config->childrenDatas_.insert(make_pair(TRIPOD_PRODUCER_CC, newConfig));
            updateFlag = true;
        }

        std::ostringstream oss1;
        oss1 << "/"<<TRIPOD_PRODUCER<<"/"<<TRIPOD_PRODUCER_JAVA;
        
        rc = zoo_exists(zk, oss1.str().c_str(), watch, 0);

        if (rc != ZOK) {
            MCE_WARN( "ConfigManager::updateProducerConfig zoo_exists failed because "<<zerror(rc)<< __FILE__<<__LINE__ );
            updateFlag =false;
            return false;
        }

        rc = zoo_get(zk, oss1.str().c_str(), watch, bc->buffer, &(bc->buflen), 0);

        if (rc != ZOK) {
            MCE_WARN( "ConfigManager::updateProducerConfig zoo_get failed because "<<zerror(rc)<< __FILE__<<__LINE__ );
            updateFlag =false;
            return false;
        }

        std::string javaVersion(bc->buffer, bc->buffer + bc->buflen);
        if(checkVersion) {
            MCE_INFO("ConfigManager::updateProducerConfig java need check version");
            std::map<std::string, ConfigDataPtr>::const_iterator findIt = config->childrenDatas_.find(TRIPOD_PRODUCER_JAVA);
            if(findIt == config->childrenDatas_.end()) {
                MCE_WARN("ConfigManager::updateProducerConfig no java");
                updateFlag =false;
                return false;
            }
            if(findIt->second->version_ != javaVersion) {
                updateFlag = true;
                findIt->second->version_ = javaVersion;
                bool sucJavaFlag = getChildrenDatas(findIt->second, zk, oss1);
                if(sucJavaFlag == false) {
                    updateFlag =false;
                    MCE_WARN("ConfigManager::updateProducerConfig get java childrendatas failed");
                    return false;
                }
            }
        }else {
            MCE_INFO("ConfigManager::updateProducerConfig java do not need check version");
            ConfigDataPtr newConfig(new ConfigData);
            newConfig->version_ = javaVersion;
            bool sucJavaFlag = getChildrenDatas(newConfig, zk, oss1);
            if(sucJavaFlag == false) {
                MCE_WARN("ConfigManager::updateProducerConfig get java childrendatas failed");
                updateFlag =false;
                return false;
            }
            config->childrenDatas_.insert(make_pair(TRIPOD_PRODUCER_JAVA, newConfig));
            updateFlag = true;
        }

    }catch(...)
    {
        MCE_WARN( "ConfigManager::updateProducerConfig can not init config "<< __FILE__<<__LINE__ );
        updateFlag =false;
        return false;
    }

    return true;
}

bool ConfigManager::updateCacheConfig(const ConfigDataPtr& config, zhandle_t* zk, bool checkVersion, bool& updateFlag) {
    MCE_INFO("ConfigManager::updateCacheConfig");

    int watch = 1;
    updateFlag =false;

    try
    {

        BufferContainerPtr bc( new BufferContainer ); 

        std::ostringstream oss;
        oss << "/"<<TRIPOD_CACHE;
        
        int rc = zoo_exists(zk, oss.str().c_str(), watch, 0);

        if (rc != ZOK) {
            MCE_WARN( "ConfigManager::updateCacheConfig zoo_exists failed because "<<zerror(rc)<< __FILE__<<__LINE__ );
            updateFlag = false;
            return false;
        }

        ConfigDataPtr childrenConfig(new ConfigData);

        bool sucChildrenFlag = getChildrenDatas(childrenConfig, zk, oss, true);

        if(sucChildrenFlag == false) {
            MCE_WARN("ConfigManager::updateCacheConfig get namespaces failed");
            updateFlag = false;
            return false;
        }

        if(checkVersion) {
            MCE_INFO("ConfigManager::updateCacheConfig need check version");
            std::map<std::string, ConfigDataPtr> oldChild;
            oldChild.swap(config->childrenDatas_);
            for(std::map<std::string, ConfigDataPtr>::const_iterator it = childrenConfig->childrenDatas_.begin(); it != childrenConfig->childrenDatas_.end(); ++it) {
                std::map<std::string, ConfigDataPtr>::const_iterator findIt = oldChild.find(it->first);
                if(findIt != oldChild.end()) {
                    if(it->second->version_ != findIt->second->version_) {
                        updateFlag = true;
                        findIt->second->version_ = it->second->version_;
                        std::ostringstream oss1;
                        oss1<<oss.str()<<"/"<<it->first;
                        bool sucChildrenFlag = getChildrenDatas(findIt->second, zk, oss1); 
                        if(sucChildrenFlag == false) {
                            MCE_WARN("ConfigManager::updateCacheConfig get namespace " << findIt->first << " children failed");
                            updateFlag = false;
                            return false;
                        }
                    }
                    config->childrenDatas_.insert(make_pair(findIt->first, findIt->second));
                } else {
                    ConfigDataPtr newConfig(new ConfigData);
                    newConfig->name_ = it->second->name_;
                    newConfig->version_ = it->second->version_;
                    
                    std::ostringstream oss1;
                    oss1<<oss.str()<<"/"<<it->first;
                    bool sucChildrenFlag = getChildrenDatas(newConfig, zk, oss1); 
                    if(sucChildrenFlag == false) {
                        MCE_WARN("ConfigManager::updateCacheConfig get namespace " << findIt->first << " children failed");
                        updateFlag = false;
                        return false;
                    }
                    config->childrenDatas_.insert(make_pair(newConfig->name_, newConfig));
                }
            }
        }else {
            MCE_INFO("ConfigManager::updateCacheConfig do not need check version");
            config->childrenDatas_.clear();
            for(std::map<std::string, ConfigDataPtr>::const_iterator it = childrenConfig->childrenDatas_.begin(); it != childrenConfig->childrenDatas_.end(); ++it) {
                std::ostringstream oss1;
                oss1<<oss.str()<<"/"<<it->first;
                
                ConfigDataPtr newConfig(new ConfigData);
                newConfig->name_ = it->second->name_;
                newConfig->version_ = it->second->version_;

                bool sucGrandsonFlag = getChildrenDatas(newConfig, zk, oss1);
                if(sucGrandsonFlag == false) {
                    MCE_WARN("ConfigManager::updateCacheConfig get namespace " << it->first << " children failed");
                    updateFlag = false;
                    return false;
                }
                config->childrenDatas_.insert(make_pair(it->first, newConfig));
            }
            updateFlag = true;
        }
    }catch(...)
    {
        MCE_WARN( "ConfigManager::updateCacheConfig can not init config "<< __FILE__<<__LINE__ );
        updateFlag = false;
        return false;
    }

    return true;
}

bool ConfigManager::getChildrenDatas(const ConfigDataPtr& config, zhandle_t* zk, const std::ostringstream& path, bool needVersion) {
    MCE_INFO("ConfigManager::getChildrenDatas path " << path.str() << " needVersion " << needVersion);
    struct String_vector nodeNames;
    int noWatch = 0;	
    int rc = zoo_get_children(zk, path.str().c_str(), noWatch, &nodeNames);

    if (rc != ZOK) {
        MCE_WARN( "ConfigManager::getChildrenDatas zoo_get_children failed because "<<zerror(rc)<< __FILE__<<__LINE__ );
        return false;
    }

    config->childrenDatas_.clear();

    for (int i = 0; i < nodeNames.count; ++i) {
        ConfigDataPtr newConfig(new ConfigData);
        newConfig->name_ = nodeNames.data[i];

        if(needVersion) {
            BufferContainerPtr bc( new BufferContainer ); 

            std::ostringstream oss;
            oss << path.str() << "/" << newConfig->name_; 

            int rc = zoo_get(zk, oss.str().c_str(), 1, bc->buffer, &(bc->buflen), 0); 
            if (rc != ZOK) {
                MCE_WARN( "ConfigManager::getChildrenDatas zoo_get failed because "<<zerror(rc)<< __FILE__<<__LINE__ );
                return false;
            } 
            newConfig->version_ = std::string(bc->buffer, bc->buffer + bc->buflen);
        }
        config->childrenDatas_.insert(make_pair(newConfig->name_, newConfig));
    }

    return true;
}

};
};
};

File: main_project/tripod-new/src/java/TripodConfig/src/main/java/com/renren/tripod/configutil/ConfigManager.java
package com.renren.tripod.configutil;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

import org.apache.zookeeper.KeeperException;
import org.apache.zookeeper.ZooKeeper;
import org.apache.zookeeper.KeeperException.NoNodeException;

public class ConfigManager {
	private static org.apache.commons.logging.Log _logger = org.apache.commons.logging.LogFactory
			.getLog(ConfigManager.class);

	private String _zkAddress = "";
	private ConfigData _config = null;
	private List<ConfigClient> _clients = new ArrayList<ConfigClient>();
	private ReadWriteLock _clientsLock = new ReentrantReadWriteLock();
	private ReadWriteLock _lock = new ReentrantReadWriteLock();

	public boolean initialize(String zkAddress) throws IOException,
			KeeperException, InterruptedException {
		_logger.info("ConfigManager.initialize zkAddress " + zkAddress);
		_zkAddress = zkAddress;
		ZkManager zm = ZkManagerFactory.get(_zkAddress);
		return updateConfig(zm.getZk());
	}

	public void registerClient(ConfigClient cc) {
		_logger.info("ConfigManager.registerClient  " + cc.toString());
		Lock wLock = _clientsLock.writeLock();
		wLock.lock();
		_clients.add(cc);
		wLock.unlock();
	}

	public boolean updateConfig(ZooKeeper zk) throws KeeperException,
			InterruptedException {
		_logger.info("ConfigManager.updateConfig");
		try {
			return init(zk);
		} catch (NoNodeException e) {
			e.printStackTrace();
			_logger.error("ConfigManager.updateConfig " + e);
			throw e;
		} catch (KeeperException e) {
			e.printStackTrace();
			_logger.error("ConfigManager.updateConfig " + e);
			throw e;
		} catch (InterruptedException e) {
			e.printStackTrace();
			_logger.error("ConfigManager.updateConfig " + e);
			throw e;
		} catch (Throwable e) {
			e.printStackTrace();
			_logger.error("ConfigManager.updateConfig " + e);
		}
		return false;
	}

	private boolean init(ZooKeeper zk) throws KeeperException,
			InterruptedException {
		_logger.info("ConfigManager.init");
		ConfigData oldConfig = getConfigData();
		if (oldConfig == null) {
			_logger.info("ConfigManager.init oldConfig == null");
			ConfigData newCacheManagerConfig = updateCacheManagerConfig(null,
					zk);
			ConfigData newProducerManagerConfig = updateProducerManagerConfig(
					null, zk);
			ConfigData newCacheConfig = updateCacheConfig(null, zk);

			ConfigData newConfig = new ConfigData();
			if (newCacheManagerConfig != null
					&& newProducerManagerConfig != null
					&& newCacheConfig != null) {
				newConfig._childrenDatas.put(ConfigData.TRIPOD_MANAGER,
						newCacheManagerConfig);
				newConfig._childrenDatas.put(ConfigData.TRIPOD_PRODUCER,
						newProducerManagerConfig);
				newConfig._childrenDatas.put(ConfigData.TRIPOD_CACHE,
						newCacheConfig);
				return setConfigData(newConfig);
			}
			_logger.error("ConfigManager.init failed");
			return false;
		}

		ConfigData oldCacheManagerConfig = oldConfig._childrenDatas
				.get(ConfigData.TRIPOD_MANAGER);
		if (oldCacheManagerConfig == null) {
			_logger.error("ConfigManager.init no cachemanager in oldConfig");
			return false;
		}
		ConfigData newCacheManagerConfig = updateCacheManagerConfig(
				oldCacheManagerConfig, zk);

		ConfigData oldProducerManagerConfig = oldConfig._childrenDatas
				.get(ConfigData.TRIPOD_PRODUCER);
		if (oldProducerManagerConfig == null) {
			_logger.error("ConfigManager.init no producermanager in oldConfig");
			return false;
		}
		ConfigData newProducerManagerConfig = updateProducerManagerConfig(
				oldProducerManagerConfig, zk);

		ConfigData oldCacheConfig = oldConfig._childrenDatas
				.get(ConfigData.TRIPOD_CACHE);
		if (oldCacheConfig == null) {
			_logger.error("ConfigManager.init no cache in oldConfig");
			return false;
		}
		ConfigData newCacheConfig = updateCacheConfig(oldCacheConfig, zk);

		boolean updateFlag = false;
		ConfigData newConfig = oldConfig;

		if (newCacheManagerConfig != null) {
			_logger.info("ConfigManager.init cachemanager has new info");
			newConfig._childrenDatas.put(ConfigData.TRIPOD_MANAGER,
					newCacheManagerConfig);
			updateFlag = true;
		}

		if (newProducerManagerConfig != null) {
			_logger.info("ConfigManager.init producermanager has new info");
			newConfig._childrenDatas.put(ConfigData.TRIPOD_PRODUCER,
					newProducerManagerConfig);
			updateFlag = true;
		}

		if (newCacheConfig != null) {
			_logger.info("ConfigManager.init cache has new info");
			newConfig._childrenDatas.put(ConfigData.TRIPOD_CACHE,
					newCacheConfig);
			updateFlag = true;
		}

		_logger.info("ConfigManager.init updateFlag " + updateFlag);
		if (updateFlag) {
			setConfigData(newConfig);
		}

		return true;
	}

	private ConfigData updateCacheManagerConfig(ConfigData oldConfig,
			ZooKeeper zk) throws KeeperException, InterruptedException {
		_logger.info("ConfigManager.updateCacheManagerConfig");
		ConfigData newConfig;
		if (oldConfig == null) {
			_logger.info("ConfigManager.updateCacheManagerConfig oldConfig == null");
			newConfig = new ConfigData();
			newConfig._name = ConfigData.TRIPOD_MANAGER;
		} else {
			newConfig = oldConfig.clone();
		}

		try {
			if (zk.exists("/" + ConfigData.TRIPOD_MANAGER, true) == null) {
				_logger.error("ConfigManager.updateCacheManagerConfig path "
						+ "/" + ConfigData.TRIPOD_MANAGER + " not in zk!");
				return null;
			}

			String version = new String(zk.getData("/"
					+ ConfigData.TRIPOD_MANAGER, true, null));

			if (oldConfig != null) {
				if (version.equals(oldConfig._version)) {
					_logger.info("ConfigManager.updateCacheManagerConfig oldConfig and newConfig hava same version");
					return null;
				}
			}

			newConfig._version = version;

			List<String> nodeNames = zk.getChildren("/"
					+ ConfigData.TRIPOD_MANAGER, false);

			newConfig._childrenDatas.clear();
			for (String nodeName : nodeNames) {
				ConfigData tConfig = new ConfigData();
				tConfig._name = nodeName;
				newConfig._childrenDatas.put(nodeName, tConfig);
			}
		} catch (NoNodeException e) {
			_logger.error("ConfigManager.updateCacheManagerConfig " + e);
			throw e;
		} catch (KeeperException e) {
			_logger.error("ConfigManager.updateCacheManagerConfig " + e);
			throw e;
		} catch (IllegalArgumentException e) {
			_logger.error("ConfigManager.updateCacheManagerConfig " + e);
			throw e;
		} catch (InterruptedException e) {
			e.printStackTrace();
			_logger.error("ConfigManager.updateCacheManagerConfig " + e);
			throw e;
		} catch (Throwable e) {
			_logger.error("ConfigManager.updateCacheManagerConfig " + e);
		}
		_logger.info("ConfigManager.updateCacheManagerConfig " + newConfig);
		return newConfig;
	}

	private ConfigData updateProducerManagerConfig(ConfigData oldConfig,
			ZooKeeper zk) throws KeeperException, InterruptedException {
		_logger.info("ConfigManager.updateProducerManagerConfig");
		ConfigData newConfig;
		if (oldConfig == null) {
			_logger.info("ConfigManager.updateProducerManagerConfig oldConfig == null");
			newConfig = new ConfigData();
			newConfig._name = ConfigData.TRIPOD_PRODUCER;
		} else {
			newConfig = oldConfig.clone();
		}

		try {

			if (zk.exists("/" + ConfigData.TRIPOD_PRODUCER + "/"
					+ ConfigData.TRIPOD_PRODUCER_CC, true) == null) {
				_logger.error("ConfigManager.updateProducerManagerConfig path "
						+ "/" + ConfigData.TRIPOD_PRODUCER + "/"
						+ ConfigData.TRIPOD_PRODUCER_CC + " not in zk!");
				return null;
			}

			if (zk.exists("/" + ConfigData.TRIPOD_PRODUCER + "/"
					+ ConfigData.TRIPOD_PRODUCER_JAVA, true) == null) {
				_logger.error("ConfigManager.updateProducerManagerConfig path "
						+ "/" + ConfigData.TRIPOD_PRODUCER + "/"
						+ ConfigData.TRIPOD_PRODUCER_JAVA + " not in zk!");
				return null;
			}

			String ccVersion = new String(zk.getData("/"
					+ ConfigData.TRIPOD_PRODUCER + "/"
					+ ConfigData.TRIPOD_PRODUCER_CC, true, null));

			String javaVersion = new String(zk.getData("/"
					+ ConfigData.TRIPOD_PRODUCER + "/"
					+ ConfigData.TRIPOD_PRODUCER_JAVA, true, null));

			boolean ccUpdateFlag = true;
			boolean javaUpdateFlag = true;
			ConfigData ccConfig = null;
			ConfigData javaConfig = null;
			if (oldConfig != null) {
				ccConfig = oldConfig._childrenDatas
						.get(ConfigData.TRIPOD_PRODUCER_CC);
				if (ccConfig == null) {
					_logger.error("ConfigManager.updateProducerManagerConfig no cc producer in oldConfig!");
					return null;
				}
				javaConfig = oldConfig._childrenDatas
						.get(ConfigData.TRIPOD_PRODUCER_JAVA);
				if (javaConfig == null) {
					_logger.error("ConfigManager.updateProducerManagerConfig no java producer in oldConfig!");
					return null;
				}
				if (ccVersion.equals(ccConfig._version)) {
					_logger.info("ConfigManager.updateProducerManagerConfig oldConfig and newConfig have same version for cc");
					ccUpdateFlag = false;
				}
				if (javaVersion.equals(javaConfig._version)) {
					_logger.info("ConfigManager.updateProducerManagerConfig oldConfig and newConfig have same version for java");
					ccUpdateFlag = false;
				}
			}

			if (!ccUpdateFlag && !javaUpdateFlag) {
				_logger.info("ConfigManager.updateProducerManagerConfig ccUpdateFlag "
						+ ccUpdateFlag + " javaUpdateFlag" + javaUpdateFlag);
				return null;
			}

			if (ccUpdateFlag) {
				_logger.info("ConfigManager.updateProducerManagerConfig ccUpdateFlag "
						+ ccUpdateFlag);
				if (ccConfig == null) {
					_logger.info("ConfigManager.updateProducerManagerConfig ccConfig == null");
					ccConfig = new ConfigData();
					ccConfig._name = ConfigData.TRIPOD_PRODUCER_CC;
				}
				ccConfig._version = ccVersion;
				ccConfig._childrenDatas.clear();

				List<String> nodeNames = zk.getChildren("/"
						+ ConfigData.TRIPOD_PRODUCER + "/"
						+ ConfigData.TRIPOD_PRODUCER_CC, false);

				for (String nodeName : nodeNames) {
					ConfigData tConfig = new ConfigData();
					tConfig._name = nodeName;
					ccConfig._childrenDatas.put(nodeName, tConfig);
				}
				newConfig._childrenDatas.put(ConfigData.TRIPOD_PRODUCER_CC,
						ccConfig);
			}

			if (javaUpdateFlag) {
				_logger.info("ConfigManager.updateProducerManagerConfig javaUpdateFlag "
						+ javaUpdateFlag);
				if (javaConfig == null) {
					_logger.info("ConfigManager.updateProducerManagerConfig javaConfig == null");
					javaConfig = new ConfigData();
					javaConfig._name = ConfigData.TRIPOD_PRODUCER_JAVA;
				}
				javaConfig._version = javaVersion;
				javaConfig._childrenDatas.clear();

				List<String> nodeNames = zk.getChildren("/"
						+ ConfigData.TRIPOD_PRODUCER + "/"
						+ ConfigData.TRIPOD_PRODUCER_JAVA, false);

				for (String nodeName : nodeNames) {
					ConfigData tConfig = new ConfigData();
					tConfig._name = nodeName;
					javaConfig._childrenDatas.put(nodeName, tConfig);
				}
				newConfig._childrenDatas.put(ConfigData.TRIPOD_PRODUCER_JAVA,
						javaConfig);
			}
		} catch (NoNodeException e) {
			_logger.error("ConfigManager.updateProducerManagerConfig " + e);
			throw e;
		} catch (KeeperException e) {
			_logger.error("ConfigManager.updateProducerManagerConfig " + e);
			throw e;
		} catch (IllegalArgumentException e) {
			_logger.error("ConfigManager.updateProducerManagerConfig " + e);
			throw e;
		} catch (InterruptedException e) {
			e.printStackTrace();
			_logger.error("ConfigManager.updateProducerManagerConfig " + e);
			throw e;
		} catch (Throwable e) {
			_logger.error("ConfigManager.updateProducerManagerConfig " + e);
		}
		_logger.info("ConfigManager.updateProducerManagerConfig " + newConfig);
		return newConfig;
	}

	private ConfigData updateCacheConfig(ConfigData oldConfig, ZooKeeper zk)
			throws KeeperException, InterruptedException {
		_logger.info("ConfigManager.updateCacheConfig");
		ConfigData newConfig = new ConfigData();
		newConfig._name = ConfigData.TRIPOD_CACHE;
		if (oldConfig == null) {
			_logger.info("ConfigManager.updateCacheConfig oldConfig == null");
		}

		try {

			boolean updateFlag = false;
			if (zk.exists("/" + ConfigData.TRIPOD_CACHE, true) == null) {
				_logger.error("ConfigManager.updateCacheConfig path " + "/"
						+ ConfigData.TRIPOD_CACHE + " not in zk!");
				return null;
			}

			List<String> nodeNames = zk.getChildren("/"
					+ ConfigData.TRIPOD_CACHE, false);

			for (String nodeName : nodeNames) {
				boolean tUpdateFlag = false;
				String version = null;
				if (oldConfig == null) {
					_logger.info("ConfigManager.updateCacheConfig oldConfig == null");
					tUpdateFlag = true;
					updateFlag = true;
				} else {
					ConfigData oldNConfig = oldConfig._childrenDatas
							.get(nodeName);
					if (oldNConfig != null) {
						version = new String(zk.getData("/"
								+ ConfigData.TRIPOD_CACHE + "/" + nodeName,
								true, null));
						if (!version.equals(oldNConfig._version)) {
							_logger.info("ConfigManager.updateCacheConfig newConfig and oldConfig hava different version");
							tUpdateFlag = true;
							updateFlag = true;
						}else {
							newConfig._childrenDatas.put(nodeName, oldNConfig);
						}
					}else {
						tUpdateFlag = true;
						updateFlag = true;
					}
				}
				if (tUpdateFlag == true) {
					_logger.info("ConfigManager.updateCacheConfig tUpdateFlag "
							+ tUpdateFlag);
					ConfigData tConfig = new ConfigData();
					tConfig._name = nodeName;
					tConfig._version = version;
					List<String> tNodeNames = zk.getChildren("/"
							+ ConfigData.TRIPOD_CACHE + "/" + nodeName, false);
					for (String tNodeName : tNodeNames) {
						ConfigData ttConfig = new ConfigData();
						ttConfig._name = tNodeName;
						tConfig._childrenDatas.put(tNodeName, ttConfig);
					}
					newConfig._childrenDatas.put(nodeName, tConfig);
				}
			}
			
			if (!updateFlag) {
				_logger.info("ConfigManager.updateCacheConfig updateFlag "
						+ updateFlag);
				return null;
			}
		} catch (NoNodeException e) {
			_logger.error("ConfigManager.updateCacheConfig " + e);
			throw e;
		} catch (KeeperException e) {
			_logger.error("ConfigManager.updateCacheConfig " + e);
			throw e;
		} catch (IllegalArgumentException e) {
			_logger.error("ConfigManager.updateCacheConfig " + e);
			throw e;
		} catch (InterruptedException e) {
			e.printStackTrace();
			_logger.error("ConfigManager.updateCacheConfig " + e);
			throw e;
		} catch (Throwable e) {
			_logger.error("ConfigManager.updateCacheConfig " + e);
		}
		_logger.info("ConfigManager.updateCacheConfig " + newConfig);
		return newConfig;
	}

	public ConfigData getConfigData() {
		_logger.info("ConfigManager.getData");
		Lock rLock = _lock.readLock();
		try {
			rLock.lock();
			if (_config != null) {
				return _config.clone();
			} else {
				_logger.error("ConfigManager.getData _config is null!");
				return null;
			}
		} catch (Exception e) {
			_logger.error("ConfigManager.getData Exception " + e);
			e.printStackTrace();
		}finally {
			rLock.unlock();
		}
		return null;
	}

	private boolean setConfigData(ConfigData config) {
		_logger.info("ConfigManager.setConfigData");
		{
			Lock wLock = _lock.writeLock();
			wLock.lock();
			_config = config;
			wLock.unlock();
		}
		Lock rLock = _clientsLock.readLock();
		rLock.lock();
		for (ConfigClient cc : _clients) {
			cc.notifyHandle();
		}
		rLock.unlock();
		return true;
	}

}
